### ISSUE TITLE
Base constructor return type 'object' is not a class or interface type.

### ISSUE THREAD

**Issue by kikar** [Microsoft/TypeScript#13975]
**TypeScript Version:**  2.2.0

Got this example from the 2.2 [announcement](https://blogs.msdn.microsoft.com/typescript/2017/02/02/announcing-typescript-2-2-rc/)

**Code**

```ts
export type Constructable = new (...args: any[]) => object;

export function Timestamped<BC extends Constructable>(Base: BC) {
    return class extends Base {
        timestamp = new Date();
    };
}

```

**Expected behavior:**
Extend my class with a timestamp.

**Actual behavior:**
Not compiling with error: "Base constructor return type 'object' is not a class or interface type."

**Comment by agubler** [comment]
Using TS 2.2.0 (Release Candidate) the follow works as expected:
```ts
type Constructable = new (...args: any[]) => object;

function Timestamped<BC extends Constructable>(Base: BC) {
    return class extends Base {
        timestamp = new Date();
    };
}

class MyClass { }

const result = new (Timestamped(MyClass))();

// prints timestamp
console.log(result.timestamp);
```

**Comment by kikar** [comment]
Tried from scratch:

- created an empty project
- copied your code in `index.ts`
- `npm i -S typescript@rc`
- `tsc index.ts`

What I get is:

> index.ts(4,26): error TS2509: Base constructor return type 'object' is not a class or interface type.

Do I need to add a tsconfig? Some special options?

**Comment by agubler** [comment]
Are you sure that `tsc` is the version you have just installed? what if you install the release candidate globally and try the same thing?

**Comment by kikar** [comment]
`npm ls -g --depth=0`

> └── typescript@2.2.0

`npm ls --depth=0`

> └── typescript@2.2.0

**Comment by mhegazy** [comment]
This is a duplicate of https://github.com/Microsoft/TypeScript/issues/13805. The fix for this issue was not included in `typescript@2.2.0`, it should be included in the next release. `typescript@next` should already have the fix.

**Comment by agubler** [comment]
Ha sorry I was actually using `next`! Apologies!

**Comment by kikar** [comment]
Why is it not going to be included in 2.2.0? Can you not make an rc1 that includes the fix?
The feature was already announced as part of 2.2.0

**Comment by mhegazy** [comment]
> Why is it not going to be included in 2.2.0? Can you not make an rc1 that includes the fix?
The feature was already announced as part of 2.2.0

The next release is `TS 2.2.1`.
